File: count_inversions/python/count_inversions.py
import operator
import os

def merge_sort(lst, compare = operator.lt):
	if len(lst) < 2:
		return lst[:],0

	middle = len(lst)/2
	left, left_inv = merge_sort(lst[:middle], compare)
	right, right_inv = merge_sort(lst[middle:], compare)

	result, split_inv = merge(left, right, compare)
	return result,(left_inv + right_inv + split_inv)

def merge(left, right, compare):
	result = []
	left_offset = 0
	right_offset = 0
	len_left = len(left)
	len_right = len(right)
	inversions_counter = 0

	while  left_offset < len_left and right_offset < len_right:
		if  compare(left[left_offset], right[right_offset]):
			result.append(left[left_offset])
			left_offset += 1
		else:
			result.append(right[right_offset])
			right_offset += 1
			#inversion -> Number of elemets remaining in the List
			inversions_counter += len(left[left_offset:])

	while left_offset < len_left:
		result.append(left[left_offset])
		left_offset += 1

	while right_offset < len_right:
		result.append(right[right_offset])
		right_offset += 1

	return result,inversions_counter

def test_algorithms ():
	sum = 0
	lst = []
	with open('test.txt') as f:
		for line in f:
			lst.append(int(line))

	ordered,inversions = merge_sort(lst)

	print "Number of inversions -> ", inversions

test_algorithms()

File: sort/mergeSort/count_inversions/go/count_inversions.go
package main

import "fmt"

func merge(a []int, b []int) ([]int, int) {
	var result = make([]int, len(a)+len(b))
	var i = 0
	var j = 0
	var num_inversions = 0

	for i < len(a) && j < len(b) {
		if a[i] <= b[j] {
			result[i+j] = a[i]
			i++
		} else {
			result[i+j] = b[j]
			j++
			//inversion -> Number of elemets remaining in the List
			num_inversions = num_inversions + len(a[i:])
		}
	}

	for i < len(a) {
		result[i+j] = a[i]
		i++
	}
	for j < len(b) {
		result[i+j] = b[j]
		j++
	}

	return result, num_inversions
}

func Mergesort(items []int) ([]int, int) {
	if len(items) < 2 {
		return items, 0
	}

	var middle = len(items) / 2

	var a, left_inv = Mergesort(items[:middle])
	var b, right_inv = Mergesort(items[middle:])
	var c, split_inv = merge(a, b)

	return c, (left_inv + right_inv + split_inv)
}

func main() {
	var _, f = Mergesort([]int{2, 3, 6, 5, 1, 4})
	fmt.Print("number of inversions: ", f, "\n")
}
